src/Bicep.Decompiler/Rewriters/ForExpressionSimplifierRewriter.cs

```diff
@@ -0,0 +1,190 @@
+    ///   for {index} in range(0, length({array}))
+    /// Where elements are being accessed via:
+    ///   {array}[{index}]
+    ///
+    /// Rewrites this syntax instead to:
+    ///   for {item} in {array}
+    /// Where elements are accessed via:
+    ///   {item}
+    /// </summary>
+    public class ForExpressionSimplifierRewriter : SyntaxRewriteVisitor
+    {
+        private readonly SemanticModel semanticModel;
+
+        public ForExpressionSimplifierRewriter(SemanticModel semanticModel)
+        {
+            this.semanticModel = semanticModel;
+        }
+
+        private static bool IsLoopIteratorExpression(SyntaxBase syntax, [NotNullWhen(true)] out VariableAccessSyntax? arrayVariable)
+        {
+            arrayVariable = null;
+
+            if (syntax is not FunctionCallSyntax rangeFunction ||
+                !LanguageConstants.IdentifierComparer.Equals(rangeFunction.Name.IdentifierName, "range") ||
+                rangeFunction.Arguments.Length != 2)
+            {
+                return false;
+            }
+
+            if (rangeFunction.Arguments[0].Expression is not IntegerLiteralSyntax startRange ||
+                startRange.Value != 0)
+            {
+                return false;
+            }
+
+            if (rangeFunction.Arguments[1].Expression is not FunctionCallSyntax lengthFunction ||
+                !LanguageConstants.IdentifierComparer.Equals(lengthFunction.Name.IdentifierName, "length") ||
+                lengthFunction.Arguments.Length != 1)
+            {
+                return false;
+            }
+
+            if (lengthFunction.Arguments[0].Expression is not VariableAccessSyntax variableAccess)
+            {
+                return false;
+            }
+
+            arrayVariable = variableAccess;
+            return true;
+        }
+
+        protected override SyntaxBase ReplaceForSyntax(ForSyntax syntax)
+        {
+            syntax = (ForSyntax)base.ReplaceForSyntax(syntax);
+
+            if (!IsLoopIteratorExpression(syntax.Expression, out var arrayVariable))
+            {
+                return syntax;
+            }
+
+            if (syntax.VariableSection is not LocalVariableSyntax indexVariable)
+            {
+                return syntax;
+            }
+
+            var arraySymbol = semanticModel.GetSymbolInfo(arrayVariable);
+            var arrayIndexSymbol = semanticModel.GetSymbolInfo(indexVariable);
+
+            if (arraySymbol is null || arrayIndexSymbol is null)
+            {
+                return syntax;
+            }
+
+            var arrayAccesses = new HashSet<ArrayAccessSyntax>();
+            var independentIndexAccesses = new HashSet<VariableAccessSyntax>();
+            CallbackVisitor.Visit(syntax, child => {
+                if (child is ArrayAccessSyntax arrayAccess)
+                {
+                    if (semanticModel.GetSymbolInfo(arrayAccess.BaseExpression) == arraySymbol && 
```

Comment:
>== [](start = 80, length = 2)

nit - `ReferenceEquals`?

Reply:
I hope you're not planning to overload the equality operators :P

Reply:
Haha no way! I just find ReferenceEquals is more explicit about the intent.

---
In reply to: [594617475](https://github.com/Azure/bicep/pull/1873#discussion_r594617475) [](ancestors = 594617475)

Reply:
Discussed over IM. `==` benefits from more type checking by the compiler. Makes sense!

---
In reply to: [594621296](https://github.com/Azure/bicep/pull/1873#discussion_r594621296) [](ancestors = 594621296,594617475)